components/ui/tabcounter/src/main/java/org/mozilla/components/ui/tabcounter/TabCounter.kt

```diff
@@ -0,0 +1,368 @@
+    }
+
+    private fun adjustTextSize(newCount: Int) {
+        val newRatio = if (newCount in TWO_DIGITS_TAB_COUNT_THRESHOLD..MAX_VISIBLE_TABS) {
+            TWO_DIGITS_SIZE_RATIO
+        } else {
+            ONE_DIGIT_SIZE_RATIO
+        }
+
+        if (newRatio != currentTextRatio) {
+            currentTextRatio = newRatio
+            text.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
+                override fun onGlobalLayout() {
+                    text.viewTreeObserver.removeOnGlobalLayoutListener(this)
+                    val sizeInPixel = (box.width * newRatio).toInt()
+                    if (sizeInPixel > 0) {
+                        // Only apply the size when we calculate a valid value.
+                        text.setTextSize(TypedValue.COMPLEX_UNIT_PX, sizeInPixel.toFloat())
+                    }
+                }
+            })
+        }
+    }
+
+    companion object {
+
+        internal const val MAX_VISIBLE_TABS = 99
+
+        internal const val SO_MANY_TABS_OPEN = "∞"
+        internal const val DEFAULT_TABS_COUNTER_TEXT = ":)"
+
+        internal const val ONE_DIGIT_SIZE_RATIO = 0.6f
+        internal const val TWO_DIGITS_SIZE_RATIO = 0.5f
+        internal const val TWO_DIGITS_TAB_COUNT_THRESHOLD = 10
+
+
+        // createBoxAnimatorSet
+        private const val ANIM_BOX_FADEOUT_FROM = 1.0f
+        private const val ANIM_BOX_FADEOUT_TO = 0.0f
+        private const val ANIM_BOX_FADEOUT_DURATION = 33L
+
+        private const val ANIM_BOX_MOVEUP1_FROM = 0.0f
+        private const val ANIM_BOX_MOVEUP1_TO = -5.3f
+        private const val ANIM_BOX_MOVEUP1_DURATION = 50L
+
+        private const val ANIM_BOX_MOVEDOWN2_FROM = -5.3f
+        private const val ANIM_BOX_MOVEDOWN2_TO = -1.0f
+        private const val ANIM_BOX_MOVEDOWN2_DURATION = 167L
+
+        private const val ANIM_BOX_FADEIN_FROM = 0.01f
+        private const val ANIM_BOX_FADEIN_TO = 1.0f
+        private const val ANIM_BOX_FADEIN_DURATION = 66L
+        private const val ANIM_BOX_MOVEDOWN3_FROM = -1.0f
+        private const val ANIM_BOX_MOVEDOWN3_TO = 2.7f
+        private const val ANIM_BOX_MOVEDOWN3_DURATION = 116L
+
+        private const val ANIM_BOX_MOVEDOWN4_FROM = 2.7f
+        private const val ANIM_BOX_MOVEDOWN4_TO = 0.0f
+        private const val ANIM_BOX_MOVEDOWN4_DURATION = 133L
+
+        private const val ANIM_BOX_SCALEUP1_FROM = 0.02f
+        private const val ANIM_BOX_SCALEUP1_TO = 1.05f
+        private const val ANIM_BOX_SCALEUP1_DURATION = 100L
+        private const val ANIM_BOX_SCALEUP1_DELAY = 16L
+
+        private const val ANIM_BOX_SCALEDOWN2_FROM = 1.05f
+        private const val ANIM_BOX_SCALEDOWN2_TO = 0.99f
+        private const val ANIM_BOX_SCALEDOWN2_DURATION = 116L
+
+        private const val ANIM_BOX_SCALEUP3_FROM = 0.99f
+        private const val ANIM_BOX_SCALEUP3_TO = 1.00f
+        private const val ANIM_BOX_SCALEUP3_DURATION = 133L
+
+
```

Comment:
\[ktlint\] WARNING: \[no\-consecutive\-blank\-lines\] Needless blank line\(s\) in column 1